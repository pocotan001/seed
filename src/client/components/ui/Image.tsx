import React from "react";
import styled from "styled-components";
import useIntersectionObserver from "../hooks/useIntersectionObserver";
import margin, { MarginProps } from "../styles/extends/margin";

interface OwnProps {
  src: string;
  lazy?: boolean;
}

type ImageProps = OwnProps &
  React.ImgHTMLAttributes<HTMLImageElement> &
  MarginProps;

const ROOT_MARGIN = "200px";
const TRANSPARENT_GIF =
  "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7";

const cache: Set<string> = new Set();
const isDataURI = (src: string) => src.indexOf("data:") === 0;

const Image: React.FC<ImageProps> = ({
  src,
  alt = "",
  lazy,
  m,
  mt,
  mr,
  mb,
  ml,
  ...rest
}) => {
  if (!lazy || cache.has(src) || isDataURI(src)) {
    return <img src={src} {...rest} />;
  }

  const el = React.useRef(null);
  const { isIntersecting } = useIntersectionObserver(el, {
    rootMargin: ROOT_MARGIN
  });

  if (isIntersecting) {
    cache.add(src);
  }

  return (
    <img ref={el} src={isIntersecting ? src : TRANSPARENT_GIF} {...rest} />
  );
};

export default styled(Image)`
  max-width: 100%;
  vertical-align: middle;
  ${margin};
`;
