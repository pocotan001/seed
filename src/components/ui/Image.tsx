import * as React from "react";
import styled from "styled-components";
import withMargin, {
  MarginProps
} from "~/components/styles/extends/withMargin";
import { isDataUri } from "~/domain/validators";
import Observer from "./Observer";

interface ImageProps
  extends MarginProps,
    React.ImgHTMLAttributes<HTMLImageElement> {
  src: string;
}

// Transparent GIF
const PLACEHOLDER_SRC =
  "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7";
const ROOT_MARGIN = "200px";

const cache: Set<string> = new Set();

const Image = styled<ImageProps>(
  ({ src, m, mt, mr, mb, ml, ...rest }) =>
    cache.has(src) || isDataUri(src) ? (
      <img src={src} {...rest} />
    ) : (
      // Lazy load
      <Observer rootMargin={ROOT_MARGIN} once>
        {isIntersecting => {
          if (isIntersecting) {
            cache.add(src);
          }

          return <img src={isIntersecting ? src : PLACEHOLDER_SRC} {...rest} />;
        }}
      </Observer>
    )
)`
  max-width: 100%;
  vertical-align: middle;
  ${withMargin};
`;

Image.defaultProps = {
  alt: ""
};

export default Image;
