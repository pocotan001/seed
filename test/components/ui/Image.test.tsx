import * as React from "react";
import * as renderer from "react-test-renderer";
import Image from "~/components/ui/Image";

describe("<Image>", () => {
  it("should render correctly", () => {
    const tree = renderer.create(<Image src="/alo" />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders as it is when data uri is set", () => {
    const tree = renderer
      .create(
        <Image src="data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs=" />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
