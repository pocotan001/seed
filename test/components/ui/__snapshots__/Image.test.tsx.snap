// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image> renders as it is when data uri is set 1`] = `
<img
  alt=""
  className="sc-bdVaJa eFtYDj"
  src="data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs="
/>
`;

exports[`<Image> should render correctly 1`] = `
<img
  alt=""
  className="sc-bdVaJa eFtYDj"
  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
/>
`;
